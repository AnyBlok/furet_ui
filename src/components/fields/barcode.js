/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import {fields} from './fields';
import VueBarcode from 'vue-barcode';

defineComponent('furet-ui-list-field-barcode', {
  template:`
    <span v-if="isHidden" />
    <barcode v-else v-bind:value="value" v-bind="config.options">
      Can't generate the barcode
    </barcode>
  `,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    components: {
      barcode: VueBarcode
    },
  },
})
fields.list.barcode = 'furet-ui-list-field-barcode'


defineComponent('furet-ui-form-field-barcode', {
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <barcode v-if="isReadonly" v-bind:value="value" v-bind="config.options">
        Can't generate the barcode
      </barcode>
      <b-input 
        v-else
        v-bind:value="value" 
        v-on:input="updateValue"
        v-bind:maxlength="config.maxlength"
        v-bind:placeholder="config.placeholder"
        icon-pack="fa"
        v-bind:icon="config.icon"
        v-bind:key="config.key"
      />
    </furet-ui-form-field-common-tooltip-field>
  `,
  extend: ['furet-ui-form-field-common'],
  prototype: {
    components: {
      barcode: VueBarcode
    },
  },
})
fields.form.barcode = 'furet-ui-form-field-barcode'
