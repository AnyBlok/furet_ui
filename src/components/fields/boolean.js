// /**
// This file is a part of the FuretUI project
// 
//    Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>
// 
// This Source Code Form is subject to the terms of the Mozilla Public License,
// v. 2.0. If a copy of the MPL was not distributed with this file,You can
// obtain one at http://mozilla.org/MPL/2.0/.
// **/
// import _ from 'underscore';
import {defineComponent} from '../factory'
import {fields} from './fields';
import { listTemplate } from './common';


defineComponent('furet-ui-list-field-boolean', {
  template: `
    <div>
      <span v-if="isHidden" />
      <b-checkbox 
        v-else
        v-model="checked"
        disabled
      />
    </div>`,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    computed: {
      checked () {
        return eval(this.value) ? true : false;
      },
    },
  },
})
fields.list.boolean = 'furet-ui-list-field-boolean'


defineComponent('furet-ui-list-field-yesno', {
  template: listTemplate,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    computed: {
      value () {
        const base = 'components.fields.yesno'
        return this.$t(eval(this.data[this.config.name] || '') ? `${base}.yes` : `${base}.no`);
      },
    },
  },
})
fields.list.yesno = 'furet-ui-list-field-yesno'

//export const FieldThumbnailBoolean = Vue.component('furet-ui-thumbnail-field-boolean', {
//    mixins: [ThumbnailMixin],
//    template: `
//        <div v-if="this.isHidden" />
//        <b-tooltip 
//            v-bind:label="getTooltip" 
//            v-bind:position="tooltipPosition"
//            v-else
//        >
//            <b-checkbox 
//                v-model="value" 
//                disabled
//            >
//                {{this.label}}
//            </b-checkbox>
//        </b-tooltip>`,
//    computed: {
//        value () {
//            return eval(this.data[this.name]) ? true : false;
//        },
//    }
//})


defineComponent('furet-ui-form-field-boolean', {
  template: `
    <furet-ui-form-field-common-tooltip
      v-bind:data="data"
      v-bind:config="config"
    >
      <b-checkbox 
        v-model="checked" 
        v-bind:disabled="isReadonly"
      >
        {{ $t(config.label) }}
      </b-checkbox>
    </furet-ui-form-field-common-tooltip>`,
  extend: ['furet-ui-form-field-common'],
  prototype: {
    computed: {
      checked: {
        get () {
          return eval(this.value) ? true : false;
        },
        set (value) {
          this.updateValue (value);
        },
      },
    },
  },
})
fields.form.boolean = 'furet-ui-form-field-boolean'
