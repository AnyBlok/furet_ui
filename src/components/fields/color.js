/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import {fields} from './fields';

defineComponent('furet-ui-list-field-color', {
  template:`
    <span v-if="isHidden" />
    <component 
      v-else-if="config.slot" 
      v-bind:is="component_template"
      v-bind:config="config"
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:value="value"
    />
    <b-input 
      v-else-if="value"
      type="color"
      v-bind:value="value" 
      disabled
    />
    <span v-else />
  `,
  extend: ['furet-ui-list-field-common'],
})
fields.list.color = 'furet-ui-list-field-color'

defineComponent('furet-ui-thumbnail-field-color', {
  template: `
    <furet-ui-thumbnail-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <component 
        v-if="config.slot" 
        v-bind:is="component_template"
        v-bind:config="config"
        v-bind:resource="resource"
        v-bind:data="data"
        v-bind:value="value"
      />
      <b-input 
        v-else-if="value"
        type="color"
        v-bind:value="value" 
        disabled
      />
    </furet-ui-thumbnail-field-common-tooltip-field>
  `,
  extend: ['furet-ui-thumbnail-field-common'],
})
fields.thumbnail.color = 'furet-ui-thumbnail-field-color'
fields.kanban.color = 'furet-ui-thumbnail-field-color'

defineComponent('furet-ui-form-field-color', {
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <div class="field has-addons">
        <p v-if="isReadonly" class="control is-expanded">
          <component 
            v-if="config.slot" 
            v-bind:is="component_template"
            v-bind:config="config"
            v-bind:resource="resource"
            v-bind:data="data"
            v-bind:value="value"
          />
          <b-input 
            v-else-if="data"
            type="color"
            v-bind:value="value" 
            disabled
          >
          </b-input>
        </p>
        <p v-else class="control">
          <b-input 
            style="width: 100px"
            v-if="data || !isReadonly"
            type="color"
            v-bind:value="value" 
            v-on:input="updateValue"
            v-bind:disabled="isReadonly"
          >
          </b-input>
        </p>
        <p class="control is-expanded">
          <b-input 
            v-if="!isReadonly"
            v-bind:value="value" 
            v-on:input="updateValue"
            v-bind:placeholder="config.placeholder"
            icon-pack="fa"
            v-bind:icon="config.icon"
            pattern="#[A-Fa-f1-9]{6}"
          />
        </p>
      </div>
    </furet-ui-form-field-common-tooltip-field>
  `,
  extend: ['furet-ui-form-field-common'],
})
fields.form.color = 'furet-ui-form-field-color'
