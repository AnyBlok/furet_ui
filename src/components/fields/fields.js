export const fields = {
  list: {},
  form: {},
  thumbnail: {},
  kanban: {},
};
export default {
    fields,
}
