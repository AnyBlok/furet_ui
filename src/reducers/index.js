/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
import global from './global';
import client from './client';
import rightmenu from './right-menu';
import leftmenu from './left-menu';
import spaces from './spaces';
import {action_manager} from './action_manager';
import actions from './actions';
import views from './views';
import data from './data';
import translate from 'counterpart';
import * as moment from 'moment';
import 'moment';


export const dispatchAll = (dispatch, datas) => {
    _.each(datas, data => {
        if (data.type) {
            switch (data.type) {
                case 'UPDATE_LOCALES':
                    _.each(data.locales, d => {
                        translate.registerTranslations(d.locale, {furetUI: d.counterpart});
                    });
                    break
                case 'SET_LOCALE':
                    translate.setLocale(data.locale)
                    moment.locale(data.locale) 
                default:
                    dispatch(data);
            }
        }
    });
}

export default {
    global,
    client,
    rightmenu,
    leftmenu,
    spaces,
    action_manager,
    actions,
    views,
    data,
}
