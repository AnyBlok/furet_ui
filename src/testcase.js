import _ from 'underscore';

export const updateGlobal = () => {
    global.document = {
        createElement: () => {
            return {
                innerHTM: () => {},
            };
        },
        body: {
            appendChild: () => {},
        }
    };
    global.window = {
        addEventListener: () => {},
        attachEvent: () => {},
    }
    global.navigator = global.navigator || {};
    global.navigator.userAgent = 'all';
    global._ = _;
}

export default {
    updateGlobal,
}
