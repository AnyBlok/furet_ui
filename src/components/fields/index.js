import { fields } from './fields';
import './common';
import './boolean';
import './datetime';
import './integer';
import './password';
import './timestamp';
import './selection';
import './string';

import './relationship';

import { defineComponent } from '../factory';

defineComponent('furet-ui-unknown-field', {
  template : `<div>Unknow Field <strong>{{config.type}}</strong></div>`,
  prototype: {
    props: ['config', 'data'],
  }
})

defineComponent('furet-ui-field', {
  template : `
    <component 
      v-bind:is="component" 
      v-bind:resource="resource" 
      v-bind:config="config" 
      v-bind:data="data"
    />
  `,
  prototype: {
    props: ['resource', 'config', 'data'],
    computed: {
      component () {
        return (fields[this.resource.type] || {})[this.config.type] || 'furet-ui-unknown-field'
      },
    },
  }
})
