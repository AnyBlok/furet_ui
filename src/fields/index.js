/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import React from 'react';
import plugin from '../plugin';
import './string';
import './selection';
import './date';
import './integer';
import './float';
import './url';
import './uuid';
import './sequence';
import './password';
import './color';
import './text';

/**
 * Unknown field, used only if the wanted field is not available
**/
plugin.set(['field'], {Unknown: (props) => {
    return <div>{props.value}</div>
}});

/**
 * return the component for viewType and fieldType
**/
export const getField = (viewType, fieldType, attribs, value) => {
    let field = plugin.get(['field', viewType, fieldType]);
    if (!field) {
        field = plugin.get(['field', 'Unknown']);
        console.log('getField', viewType, fieldType, attribs, value)
    }
    const id = 'view-type-' + viewType + '-field_type-' + fieldType + '-field-name-' + (attribs.name || '');
    return React.createElement(field, Object.assign({id}, attribs, {value}));
}

export default {
    getField,
}
