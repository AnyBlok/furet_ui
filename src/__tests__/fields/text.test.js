/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
jest.mock('react-rte')
import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import chai from 'chai';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';
import reducers from '../../reducers';
import {updateGlobal} from '../../testcase';
import {getField} from '../../fields';

test('getField for List', () => {
    const store = createStore(combineReducers(reducers));
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('List', 'Text', {}, 'Test')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Thumbnail', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('Thumbnail', 'Text', {}, 'Test')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Form', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('Form', 'Text', {}, 'Test')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Form readonly', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('Form', 'Text', {readonly: true, label: 'Test'}, 'Test')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Form required', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('Form', 'Text', {required: '1', label: 'Test'}, 'Test')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Form required with empty value', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                {getField('Form', 'Text', {required: '1', label: 'Test'}, '')}
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
