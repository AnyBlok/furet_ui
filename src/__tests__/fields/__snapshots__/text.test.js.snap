// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String Form component Render 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render invisible condition is false 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render invisible condition is true 1`] = `<div data-server-rendered="true"></div>`;

exports[`String Form component Render invisible is True 1`] = `<div data-server-rendered="true"></div>`;

exports[`String Form component Render invisible is false 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render label 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><label class="label">The label</label> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render mode readonly by config 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small"><p>The text</p></div> <!----></div></span>`;

exports[`String Form component Render mode readonly conditional param 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small"><p>The text</p></div> <!----></div></span>`;

exports[`String Form component Render mode readonly param 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small"><p>The text</p></div> <!----></div></span>`;

exports[`String Form component Render placeholder 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render required 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render required conditionnal 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render required without value 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <p class="help is-danger">this field is required</p></div></span>`;

exports[`String Form component Render tooltip left 1`] = `<span data-label="The left tooltip" data-server-rendered="true" class="tooltip is-primary is-left is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render tooltip up 1`] = `<span data-label="The tooltip" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Form component Render without data 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="quillWrapper"><div id="quill-container"></div></div> <!----></div></span>`;

exports[`String Thumbnail component Render 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render invisible condition is false 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render invisible condition is true 1`] = `<div data-server-rendered="true"></div>`;

exports[`String Thumbnail component Render invisible is True 1`] = `<div data-server-rendered="true"></div>`;

exports[`String Thumbnail component Render invisible is false 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render label 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><label class="label">The label</label> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render tooltip left 1`] = `<span data-label="The left tooltip" data-server-rendered="true" class="tooltip is-primary is-left is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render tooltip up 1`] = `<span data-label="The tooltip" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"><p>The text</p></div> <!----></div></span>`;

exports[`String Thumbnail component Render without data 1`] = `<span data-label="" data-server-rendered="true" class="tooltip is-primary is-top is-medium" style="width:100%;"><div class="field" style="width:inherit;"><!----> <div class="box content is-small" style="width:100%;"></div> <!----></div></span>`;

exports[`String list component Render invisible condition is false 1`] = `
<div data-server-rendered="true" class="content is-small" style="width:100%;">
    <p>The text</p>
</div>
`;

exports[`String list component Render invisible condition is true 1`] = `<span data-server-rendered="true"></span>`;

exports[`String list component Render invisible is True 1`] = `<span data-server-rendered="true"></span>`;

exports[`String list component Render invisible is false 1`] = `
<div data-server-rendered="true" class="content is-small" style="width:100%;">
    <p>The text</p>
</div>
`;

exports[`String list component Render with data 1`] = `
<div data-server-rendered="true" class="content is-small" style="width:100%;">
    <p>The text</p>
</div>
`;

exports[`String list component Render without data 1`] = `<div data-server-rendered="true" class="content is-small" style="width:100%;"></div>`;
